File: Excel/517 Challenge.R
library(tidyverse)
library(readxl)
library(combinat)

path = "Excel/517 Arrange Numbers to Form Square Chains.xlsx"
input = read_excel(path, range = "A1:A10") %>% unlist()
test  = read_excel(path, range = "B1:B10") %>% unlist()

is_perfect_square <- function(x) {
  sqrt_x <- sqrt(x)
  sqrt_x == floor(sqrt_x)
}

is_valid_sequence <- function(nums) {
  all(map2_lgl(nums[-length(nums)], nums[-1], ~ is_perfect_square(.x + .y)))
}

find_valid_permutation <- function(nums) {
  permutations <- permn(nums)
  valid_perm <- keep(permutations, is_valid_sequence)
  if (length(valid_perm) > 0) {
    return(valid_perm[[1]])
  } else {
    return(NULL)
  }
}

result = find_valid_permutation(input)
all.equal(unname(result), unname(test))
# [1] TRUE

File: Excel/500-599/517/517 Challenge.py
import pandas as pd
import math
import itertools

path = "517 Arrange Numbers to Form Square Chains.xlsx"
input = pd.read_excel(path, usecols="A").values.flatten()
test = pd.read_excel(path, usecols="B").values.flatten()

def is_perfect_square(x):
    return math.isqrt(x)**2 == x

def is_valid_sequence(nums):
    return all(is_perfect_square(nums[i] + nums[i+1]) for i in range(len(nums) - 1))

def find_valid_permutation(nums):
    for perm in itertools.permutations(nums):
        if is_valid_sequence(perm):
            return list(perm)
    return None

result = find_valid_permutation(input)
result = " ".join(map(str, result[::-1]))
test = " ".join(map(str, test))

print(result == test)   # True